[Model] Restore Gemma3 GGUF multimodal support with GGUF-only guards

lucianommartins opened the pull request:
## Summary

This PR restores custom attention mask generation for Gemma3 GGUF multimodal models that was partially reverted in #28995. The implementation uses robust GGUF-only file format guards to ensure the feature exclusively applies to GGUF models and does not affect HuggingFace models.

**Resolves:** #28995 (HF model regression)  
**Restores functionality from:** #27772

## Background

PR #27772 initially added Gemma3 GGUF multimodal support, enabling users to run quantized Gemma3 multimodal models with both text-only and image+text prompts. However, it was partially reverted in #28995 because the custom attention mask logic incorrectly triggered for HuggingFace models, causing test failures.

**Root cause of #28995:** The original implementation lacked file format guards, causing the custom attention mask generation to activate for both GGUF and HF models.

## Solution

This PR addresses the regression by implementing a **3-layer defense-in-depth guard mechanism**:

### Layer 1: Model Format Check (Primary Guard)
```python
def uses_custom_attention_masks(config: PretrainedConfig, model_path: str) -> bool:
    """Only return True for GGUF Gemma3 multimodal models."""
    architectures = getattr(config, "architectures", [])
    is_gemma3 = "Gemma3ForConditionalGeneration" in architectures
    is_gguf = check_gguf_file(model_path)  # ← Critical GGUF guard
    return is_gemma3 and is_gguf
```

### Layer 2: Multimodal Feature Check
```python
has_mm_features = any(
    req_state.mm_features for req_state in self.requests.values()
)
```

### Layer 3: Method Existence Check
```python
hasattr(self.model, "generate_attention_masks")
```

**Result:** HF models never have `uses_custom_attention_masks = True`, preventing the issue that caused #28995.

## Changes

### Files Modified (4)

1. **`vllm/transformers_utils/config.py`**
   - Add `uses_custom_attention_masks()` utility function
   - Implements GGUF file format check using `check_gguf_file()`

2. **`vllm/config/model.py`**
   - Add `uses_custom_attention_masks` property to `ModelConfig`
   - Delegates to utility function with model path for GGUF detection

3. **`vllm/v1/worker/gpu_model_runner.py`**
   - Initialize `uses_custom_attention_masks` attribute in `GPUModelRunner`
   - Apply 3-layer guard before calling custom attention mask generation

4. **`vllm/model_executor/models/gemma3_mm.py`**
   - Restore `generate_attention_masks()` method
   - Generates custom masks enabling bidirectional attention between image tokens
   - Handles sliding window attention for GGUF compatibility

## Test Plan

### GGUF Model Validation

Tested with multiple quantized Gemma3 GGUF models to ensure functionality across different model sizes:

**Text-Only Inference:**
- Gemma3 1B GGUF (Q4_0 quantization)
- Gemma3 4B GGUF (Q4_0 quantization)
- Gemma3 270M GGUF (Q4_0 quantization)

**Multimodal Inference:**
- Gemma3 4B GGUF with `mmproj.gguf` vision tower
- Input: Multi-image prompts (2 images: kitten photo + autumn forest scene)
- Expected: Accurate image descriptions with proper bidirectional attention between image tokens

### HuggingFace Model Regression Testing

Executed the full vLLM multimodal test suite to verify zero impact on HF models:

```bash
pytest -s -v tests/models/multimodal/generation/test_common.py -k "gemma3-test"
```

This ensures the GGUF guards prevent any unintended activation of custom attention mask logic for HuggingFace models.

## Test Results

### GGUF Model Results (All Pass)

| Model | Quantization | Test Type | Status | Details |
|-------|-------------|-----------|--------|---------|
| Gemma3-1B | Q4_0 | Text-only | PASSED | Generates coherent haiku about coding |
| Gemma3-4B | Q4_0 | Text-only | PASSED | Generates coherent haiku about coding |
| Gemma3-270M | Q4_0 | Text-only | PASSED | Generates coherent haiku about coding |
| Gemma3-4B | Q4_0 | **Multimodal** | **PASSED** | Correctly describes both images with accurate details |

**Multimodal Output Example:**
```
Image 1: A close-up shot of a tabby kitten with striking blue eyes. The kitten 
is lying on a green surface, likely a rug or carpet. Its fur is a mix of brown 
and black stripes, and it has a slightly melancholic expression...

Image 2: A breathtaking landscape photograph of a dense pine forest bathed in 
the warm, golden light of a late autumn or early winter day...
```

### HuggingFace Model Regression Test (All Pass)

```bash
pytest -s -v tests/models/multimodal/generation/test_common.py -k gemma3-test
# Result: 8 passed, 335 deselected, 23 warnings in 915.69s (15m 15s)
```

**Test Coverage:**
- Single-image inference (3 test cases)
- Multi-image inference (3 test cases)
- Various prompt templates (2 test cases)
- **All 8 test cases pass** - confirms zero regression for HF models

### Verification of Fix for #28995

The failing test from #28995 (`pytest gemma3-test`) now **passes completely**:

**Why it works now:**
- `uses_custom_attention_masks` returns `False` for HF models (no `.gguf` file detected)
- Custom attention mask generation never executes for HF models
- HF models use their native attention mechanism without interference

## Isolation & Safety Guarantees

### How HF Models Are Protected:

1. **File Format Check:**
   ```python
   is_gguf = check_gguf_file(model_path)  # Returns False for HF models
   ```

2. **Short-Circuit Logic:**
   ```python
   return is_gemma3 and is_gguf  # Requires BOTH conditions
   ```

3. **Runtime Guard:**
   ```python
   if self.uses_custom_attention_masks:  # Always False for HF
       # Custom mask generation (never executed for HF)
   ```

### What Changed from #27772:

| Aspect | Original PR #27772 | This PR (Fixed) |
|--------|-------------------|-----------------|
| **Guard Mechanism** | Architecture check only | Architecture + GGUF file format check |
| **HF Model Impact** | Incorrectly triggered | Never triggers |
| **GGUF Multimodal** | Works | Works |

## Code Quality

- All linting checks pass (`ruff check`, `ruff format`, `mypy`)
- All pre-commit hooks pass
- Follows Google Python style guide
- Comprehensive docstrings with clear GGUF-only notes
- Defense-in-depth guard pattern

## Backward Compatibility

- **Zero impact on existing HuggingFace models** (verified via pytest)
- **Zero impact on other model architectures** (GGUF check prevents any non-Gemma3 activation)
- **Restores functionality from #27772** for GGUF users
- No API changes, no breaking changes

## Documentation

No user-facing documentation changes required. The feature is transparent to users - GGUF Gemma3 multimodal models work automatically without configuration.

## Release Notes

This fix should be included in release notes as:

> **[Model] Restored Gemma3 GGUF multimodal support** - Re-enables custom attention mask generation for Gemma3 GGUF multimodal models with robust GGUF-only guards, fixing the regression introduced in #28995 while maintaining zero impact on HuggingFace models.

---

## Checklist

- [x] PR title follows format: `[Model] <description>`
- [x] Code passes all linting checks
- [x] Comprehensive tests included (5 tests covering GGUF + HF compatibility)
- [x] Test results documented above
- [x] Code includes appropriate docstrings
- [x] Backward compatibility verified
- [x] Sign-off included (`git commit -s`)
- [x] Fixes reported issue #28995
- [x] Restores functionality from #27772

---

**Related PRs:**
- Original implementation: #27772
- Revert of custom attention masks: #28995 (this PR fixes the root cause)

lucianommartins committed:
[Model] Restore Gemma3 GGUF multimodal support with GGUF-only guards

Restores custom attention mask generation for Gemma3 GGUF multimodal models
that was partially reverted in #28995. Implements robust GGUF-only guards to
ensure the feature only applies to GGUF models and does not affect HF models.

Changes:
- Add uses_custom_attention_masks() utility with GGUF file format check
- Add uses_custom_attention_masks property to ModelConfig
- Initialize uses_custom_attention_masks in GPUModelRunner
- Restore generate_attention_masks() method to Gemma3ForConditionalGeneration
- Implement 3-layer defense-in-depth guard mechanism

The implementation uses check_gguf_file() to guarantee that custom attention
mask logic only triggers for GGUF files, preventing the issue that caused
the original revert where HF models incorrectly triggered the custom logic.

Tested with GGUF models (1B, 4B, 270M) for both text-only and multimodal
inference. HF model compatibility verified via pytest multimodal test suite.

Signed-off-by: Luciano Martins <lucianommartins@users.noreply.github.com>

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2336,6 +2337,26 @@ def _preprocess(
+                    self.positions.gpu[:num_scheduled_tokens],
+                    mask_dtype=self.model.dtype,
+                )
+                model_kwargs.update(mask_kwargs)
```
This shows that the application of attention mask is actually incorrect then.

DarkLight1337 replied:
https://github.com/vllm-project/vllm/pull/27938 should solve this right?

Isotr0py replied:
Yes

lucianommartins committed:
Fix GGUF pan-and-scan attention and CUDA graph mask preservation

Fixes seven critical issues in GGUF multimodal inference:

1. Attention scaling parameter bug (gemma3.py):
   - Fix F.scaled_dot_product_attention to use named parameters
   - Changed positional args to attn_mask=attn_mask, scale=self.scaling
   - Prevents incorrect dropout application (was 6.25% instead of 0%)

2. Custom attention mask persistence (gpu_model_runner.py):
   - Store custom_model_kwargs after mask generation
   - Merge custom_model_kwargs in _dummy_run
   - Prevents loss of attention masks during CUDA graph re-initialization

3. Pan-and-scan attention pattern (gemma3_mm.py):
   - Detect pan-and-scan mode via multimodal_config.do_pan_and_scan
   - Prevents crop isolation artifacts in sequential processing

4. GGUF unquantized weight loading (weight_utils.py):
   - Add proper dtype conversion for BF16/F16/F32 stored as uint8
   - Handle byte-to-dtype conversion (BF16: 2 bytes, F16: 2 bytes, F32: 4 bytes)
   - Add fallback handling for unexpected dtype/type combinations
   - Fixes weight loading for unquantized GGUF multimodal projector weights

5. Fix GGUF pan-and-scan attention and token handling:
   - Unified bidirectional attention for all image tokens (fixes random output)
   - Fixed pan-and-scan token splits via position discontinuity detection
   - Corrected num_image_patches to reflect actual crop count from pixel_values

6. Add GGUF multimodal integration tests (test_multimodal_gguf.py):
   - Compare GGUF (vLLM) output against native HuggingFace (HfRunner)
   - Test regular multimodal inference with QAT Q4_0 GGUF
   - Test pan-and-scan multimodal with unquantized BF16 GGUF
   - Use @create_new_process_for_each_test for subprocess isolation

7. Robust sequence detection for batched multimodal (gemma3_mm.py):
   - Replace  heuristic with
   - Pass query_start_loc from gpu_model_runner.py to generate_attention_masks()
   - Fixes multi-scale batching failures where num_computed_tokens > 0
   - Ensures correct attention mask generation for all batch configurations

Signed-off-by: Luciano Martins <lucianommartins@users.noreply.github.com>

lucianommartins commented:
done, @Isotr0py - now the whole `test_multimodal_gguf.py` tests the GGUF models against HF ones using `HFRunner`.

lucianommartins commented:
@Isotr0py - do you know the memory limits of the containers running the tests? my multimodal test failed with segment fault when trying to load the `gemma-3-4b-it` for `HFRunner`:

```
Processed prompts:   0% 0/2 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]
Processed prompts:  50% 1/2 [00:02<00:02,  2.46s/it, est. speed input: 114.04 toks/s, output: 12.99 toks/s]
Processed prompts: 100% 2/2 [00:02<00:00,  2.46s/it, est. speed input: 220.28 toks/s, output: 25.40 toks/s]
Processed prompts: 100% 2/2 [00:02<00:00,  1.26s/it, est. speed input: 220.28 toks/s, output: 25.40 toks/s]
[2025-11-28T20:58:29Z] [rank0]:[W1128 12:58:29.763342309 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[2025-11-28T20:58:31Z] Fatal Python error: Segmentation fault
```

with the stack pointing to:

```
File "/vllm-workspace/tests/models/multimodal/generation/test_multimodal_gguf.py", line 121 in run_multimodal_gguf_test
File "/vllm-workspace/tests/conftest.py", line 355 in _init
File "/vllm-workspace/tests/conftest.py", line 293 in __init__
File "/usr/local/lib/python3.12/dist-packages/transformers/models/gemma3/modeling_gemma3.py", line 998 in __init__
```

(line 121 from `test_multimodal_gguf.py` is `with hf_runner(...)`...

Also it is suspicious that I have a larger GPU on my test environment and this test runned with no errors:

```
$ pytest tests/models/multimodal/generation/test_multimodal_gguf.py -v -s
========================================================== test session starts ===========================================================
platform linux -- Python 3.13.7, pytest-9.0.1, pluggy-1.6.0 -- /projects/gemma-vllm/fixes/P0.1b_Multimodal_GGUF/bugfix/mm-bugfix-env/bin/python3
cachedir: .pytest_cache
rootdir: /projects/gemma-vllm/fixes/P0.1b_Multimodal_GGUF/bugfix/my-vllm
configfile: pyproject.toml
plugins: anyio-4.11.0
collected 2 items                                                                                                

<...>

============================================================ warnings summary ============================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================== 2 passed, 2 warnings in 525.68s (0:08:45) ================================================
```

lucianommartins commented:
hey @Isotr0py - It is all set now. the GGUF multimodal tests are being triggered and passing without segmentation fault.

let's merge it? :)

lucianommartins commented:
Hi @Isotr0py / @sgerrard -

I've done some extensive GGUF testing using https://github.com/vllm-project/vllm/pull/27938 and:
- text inferences (both single and in batches) work perfectly for both quantized and unquantized models
- single image inferences (both single and in batches) work perfectly for both quantized and unquantized models
- multi image inferences (both single and in batches) work perfectly for both quantized and unquantized models

but pan and scan is still suboptimal. while the HF model gives:

```
Generated text:
Here's a breakdown of the content in the center of the image, based on the provided crops:

*   **Chinese Archway:** A prominent red and gold Chinese archway dominates the center. It has traditional Chinese characters (likely "Temple") inscribed on it.
*   **Lion Statues:** Two white lion statues flank the archway, adding to the cultural theme.
*   **Street:** A street runs directly beneath the archway, with a car driving along it.
*   **Buildings:** Buildings with shops and businesses are visible on either side of the street, partially obscured by the archway.
```

the same unquantized model using the GGUF variant only gives (probably a problem on how the crops are being injected into the prompt and how the attention process is working for both the original image and the crops):

```
Generated text:
The content in the center of the image is a bright red stop sign.
```

@Isotr0py - I also want to make a sync with you. that PR is amazing!! but I would like to encompass our efforts to avoid lost cycles (like the one for this PR) as also bringing more cool things into VLLM faster.

I will close this PR and raise an issue just for the record of this pan and scan limitation.
